### DISCUSSION TITLE
Emscripten and Wasm GC proposal

### DISCUSSION THREAD

**Discussion by alienself** [emscripten-core/emscripten#19227] [General]
Hi, 

I was wondering how will WASM GC impact Emscripten (https://github.com/WebAssembly/gc/issues)

Will it be possible for C++ to share the Javaascript heap and call web api directly or will this be reserved for higher level languages (such as Java or Dart)?

**Comment by sbc100** [comment]
As it stands, Wasm GC doesn't do much to make it easier to call Web APIs since Wasm GC objects cannot, in general, be used as arguments to Web APIs, or any JS APIs for that matter.   With the MVP, Wasm GC objects appear to JS as opaque references, without any direct field access.    Perhaps as a followup, post-MVP there will be better integration added.

In other words, even with Wasm GC you still still need some kind of shim layer to convert Wasm GC objects into JS objects.

Looking forward, even in a world where (somehow) Wasm GC object can be passed as argument to JS APIs its unlikely that Wasm GC objects will have good integration into C/C++.  There are some plans afoot, but they are still a fair way off: https://github.com/Igalia/ref-cpp

**Reply by alienself** [reply]
Thanks a lot for this explanation and the ref-cpp project link.  
Same here, I wish we could completely get rid of the js layer and call web api directly or at the very least reduce the need for marshalling. 

So overall since c++ doesn't need GC in the first place, do you think that emscripten will not directly support it? 

Would it be possible to use Wasm GC similar to what Unreal is doing as a totally optional thing? https://unrealcommunity.wiki/garbage-collection-36d1da

Related: Have you watched this recent talk about Wasm GC? https://www.youtube.com/watch?v=fDNXomIn53Y
![Screenshot 2023-04-21 152931](https://user-images.githubusercontent.com/87069099/233743851-e9b32246-65d6-4471-bb63-97914751e9f2.png)

**Reply by sbc100** [reply]
Emscripten will likely support GC object via assembly, but its unlikely to support it in the C/C++ language.

The GC you are referring to in Unreal is userspace GC in linear memory, and not connected to Wasm GC or the JavaScript engine's GC. One of the reasons the Wasm GC proposal exists is to try to avoid such things, but you could do that today with emscripten by building a GC into your project (e.g. boehm GC: https://en.wikipedia.org/wiki/Boehm_garbage_collector).

I did watch that great talk from wasm I/O.  I think was perhaps a little forward looking with respect being able to directly talk to web APIs.. perhaps envisaging the post-MVP world?
